### PR TITLE
Full 1.13 support + 2.16.0 release

### PR DESCRIPTION
**Pull request by mdcfe**
See the [1.13 milestone](https://github.com/EssentialsX/Essentials/milestone/3) and #2223 for progress.

### CODE REVIEW

File: Essentials/src/com/earth2me/essentials/MetaItemStack.java [hunk 1]
```diff
@@ -212,14 +218,14 @@ public void addStringMeta(final CommandSource sender, final boolean allowUnsafe,
             final BookMeta meta = (BookMeta) stack.getItemMeta();
             meta.setTitle(title);
             stack.setItemMeta(meta);
-        } else if (split.length > 1 && split[0].equalsIgnoreCase("power") && stack.getType() == Material.FIREWORK && hasMetaPermission(sender, "firework-power", false, true, ess)) {
+        } else if (split.length > 1 && split[0].equalsIgnoreCase("power") && (stack.getType() == Material.FIREWORK_ROCKET || stack.getType() == Material.FIREWORK_STAR)&& hasMetaPermission(sender, "firework-power", false, true, ess)) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/MetaItemStack.java]: Refactor to use MaterialUtil.

File: Essentials/src/com/earth2me/essentials/MobData.java [hunk 2]
```diff
@@ -53,9 +53,9 @@
     BAY_HORSE("bay", EntityType.HORSE, Horse.Color.BROWN, true),
     BROWN_HORSE("brown", EntityType.HORSE, Horse.Color.BROWN, false),
     SADDLE_HORSE("saddle", EntityType.HORSE, Data.HORSESADDLE, true),
-    GOLD_ARMOR_HORSE("goldarmor", EntityType.HORSE, Material.GOLD_BARDING, true),
-    DIAMOND_ARMOR_HORSE("diamondarmor", EntityType.HORSE, Material.DIAMOND_BARDING, true),
-    ARMOR_HORSE("armor", EntityType.HORSE, Material.IRON_BARDING, true),
+    GOLD_ARMOR_HORSE("goldarmor", EntityType.HORSE, Material.GOLDEN_HORSE_ARMOR, true),
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/MobData.java]: Use MaterialUtil for backwards-compatibility.

File: Essentials/src/com/earth2me/essentials/SpawnMob.java [hunk 3]
```diff
@@ -191,10 +191,10 @@ private static void changeMobData(final CommandSource sender, final EntityType t
                     invent.setChestplate(new ItemStack(Material.DIAMOND_CHESTPLATE, 1));
                     invent.setHelmet(new ItemStack(Material.DIAMOND_HELMET, 1));
                 } else if (inputData.contains("gold")) {
-                    invent.setBoots(new ItemStack(Material.GOLD_BOOTS, 1));
-                    invent.setLeggings(new ItemStack(Material.GOLD_LEGGINGS, 1));
-                    invent.setChestplate(new ItemStack(Material.GOLD_CHESTPLATE, 1));
-                    invent.setHelmet(new ItemStack(Material.GOLD_HELMET, 1));
+                    invent.setBoots(new ItemStack(Material.GOLDEN_BOOTS, 1));
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/SpawnMob.java]: MaterialUtil here too.

File: Essentials/src/com/earth2me/essentials/commands/Commandbook.java [hunk 4]
```diff
@@ -44,15 +44,15 @@ public void run(final Server server, final User user, final String commandLabel,
                 }
             } else {
                 if (isAuthor(bmeta, player) || user.isAuthorized("essentials.book.others")) {
-                    ItemStack newItem = new ItemStack(Material.BOOK_AND_QUILL, item.getAmount());
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandbook.java:47]: MaterialUtil

File: Essentials/src/com/earth2me/essentials/commands/Commandfirework.java [hunk 5]
```diff
@@ -43,8 +43,8 @@ public Commandfirework() {
 
     @Override
     protected void run(final Server server, final User user, final String commandLabel, final String[] args) throws Exception {
-        final ItemStack stack = user.getBase().getItemInHand();
-        if (stack.getType() == Material.FIREWORK) {
+        final ItemStack stack = user.getBase().getInventory().getItemInMainHand();
+        if (stack.getType() == Material.FIREWORK_ROCKET || stack.getType() == Material.FIREWORK_STAR) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandfirework.java]: MaterialUtil again

File: Essentials/src/com/earth2me/essentials/commands/Commanditemdb.java [hunk 6]
```diff
@@ -21,17 +21,18 @@ protected void run(Server server, CommandSource sender, String commandLabel, Str
...
-                itemStack = sender.getPlayer().getItemInHand();
+                itemStack = sender.getPlayer().getInventory().getItemInMainHand();
             }
             if (itemStack == null) {
                 throw new NotEnoughArgumentsException();
             }
         } else {
             itemStack = ess.getItemDb().get(args[0]);
         }
-        sender.sendMessage(tl("itemType", itemStack.getType().toString(), itemStack.getTypeId() + ":" + Integer.toString(itemStack.getDurability())));
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commanditemdb.java:34]: We should re-add the item ID message when using 1.12 or below.

File: Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java [hunk 7]
```diff
@@ -64,11 +64,11 @@ public void run(final Server server, final CommandSource sender, final String co
         } else if (selectedRecipe instanceof ShapedRecipe) {
             shapedRecipe(sender, (ShapedRecipe) selectedRecipe, sender.isPlayer());
         } else if (selectedRecipe instanceof ShapelessRecipe) {
-            if (recipesOfType.size() == 1 && itemType.getType() == Material.FIREWORK) {
+            if (recipesOfType.size() == 1 && (itemType.getType() == Material.FIREWORK_ROCKET || itemType.getType() == Material.FIREWORK_STAR)) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandrecipe.java]: MaterialUtil

File: Essentials/src/com/earth2me/essentials/commands/Commandskull.java [hunk 8]
```diff
@@ -31,14 +31,14 @@ protected void run(final Server server, final User user, final String commandLab
             owner = user.getName();
         }
 
-        ItemStack itemSkull = user.getBase().getItemInHand();
+        ItemStack itemSkull = user.getBase().getInventory().getItemInMainHand();
         SkullMeta metaSkull = null;
         boolean spawn = false;
 
-        if (itemSkull != null && itemSkull.getType() == Material.SKULL_ITEM && itemSkull.getDurability() == 3) {
+        if (itemSkull != null && itemSkull.getType() == Material.LEGACY_SKULL_ITEM && itemSkull.getDurability() == 3) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandskull.java]: MaterialUtil

File: Essentials/src/com/earth2me/essentials/commands/Commandwhois.java [hunk 9]
```diff
@@ -16,8 +17,13 @@
 
 
 public class Commandwhois extends EssentialsCommand {
+    private final Statistic playOneTick;
+
     public Commandwhois() {
         super("whois");
+        // For some reason, in 1.13 PLAY_ONE_MINUTE = ticks played
+        // https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/b848d8ce633871b52115247b089029749c02f579
+        playOneTick = EnumUtil.getStatistic("PLAY_ONE_MINUTE", "PLAY_ONE_TICK");
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/commands/Commandwhois.java]: Note to past/future MD: why isn't this static?

File: Essentials/src/com/earth2me/essentials/craftbukkit/FakeWorld.java [hunk 10]
```diff
@@ -164,8 +163,8 @@ public boolean generateTree(Location lctn, TreeType tt) {
     }
 
     @Override
-    public boolean generateTree(Location lctn, TreeType tt, BlockChangeDelegate bcd) {
-        throw new UnsupportedOperationException("Not supported yet.");
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/craftbukkit/FakeWorld.java]: Why was this changed?

File: Essentials/src/com/earth2me/essentials/items/AbstractItemDb.java [hunk 11]
```diff
@@ -341,7 +144,8 @@ public String serialize(ItemStack is) {
                     sb.append(e.getName().toLowerCase()).append(":").append(enchantmentStorageMeta.getStoredEnchantLevel(e)).append(" ");
                 }
                 break;
-            case FIREWORK:
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/items/AbstractItemDb.java]: This is not 1.12-friendly.

File: Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java [hunk 12]
```diff
@@ -213,7 +213,7 @@ protected boolean onBlockPush(final Block block, final IEssentials ess) {
 
     protected static boolean checkIfBlockBreaksSigns(final Block block) {
         final Block sign = block.getRelative(BlockFace.UP);
-        if (sign.getType() == Material.SIGN_POST && isValidSign(new BlockSign(sign))) {
+        if (sign.getType() == Material.SIGN && isValidSign(new BlockSign(sign))) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java]: MaterialUtil

File: Essentials/src/com/earth2me/essentials/signs/SignBlockListener.java [hunk 13]
```diff
@@ -22,7 +22,7 @@
 public class SignBlockListener implements Listener {
     private static final Logger LOGGER = Logger.getLogger("Essentials");
     private static final Material WALL_SIGN = Material.WALL_SIGN;
-    private static final Material SIGN_POST = Material.SIGN_POST;
+    private static final Material SIGN_POST = Material.SIGN;
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignBlockListener.java]: MaterialUtil

File: Essentials/src/com/earth2me/essentials/signs/SignEntityListener.java [hunk 14]
```diff
@@ -25,7 +25,7 @@ public void onSignEntityExplode(final EntityExplodeEvent event) {
         }
 
         for (Block block : event.blockList()) {
-            if (((block.getType() == Material.WALL_SIGN || block.getType() == Material.SIGN_POST) && EssentialsSign.isValidSign(ess, new EssentialsSign.BlockSign(block))) || EssentialsSign.checkIfBlockBreaksSigns(block)) {
+            if (((block.getType() == Material.WALL_SIGN || block.getType() == Material.SIGN) && EssentialsSign.isValidSign(ess, new EssentialsSign.BlockSign(block))) || EssentialsSign.checkIfBlockBreaksSigns(block)) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignEntityListener.java]: MaterialUtil

File: Essentials/src/com/earth2me/essentials/signs/SignPlayerListener.java [hunk 15]
```diff
@@ -54,7 +54,7 @@ public void onSignPlayerInteract(final PlayerInteractEvent event) {
         }
 
         final Material mat = block.getType();
-        if (mat == Material.SIGN_POST || mat == Material.WALL_SIGN) {
+        if (mat == Material.SIGN || mat == Material.WALL_SIGN) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignPlayerListener.java]: MaterialUtil

File: Essentials/src/com/earth2me/essentials/signs/SignProtection.java [hunk 16]
```diff
@@ -26,7 +26,6 @@
     public SignProtection() {
         super("Protection");
         protectedBlocks.add(Material.CHEST);
-        protectedBlocks.add(Material.BURNING_FURNACE);
         protectedBlocks.add(Material.FURNACE);
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignProtection.java]: Perhaps add BURNING_FURNACE in a try/catch for backwards-compatibility?

File: Essentials/src/com/earth2me/essentials/signs/SignProtection.java [hunk 17]
```diff
@@ -103,7 +102,7 @@ private void getConnectedSigns(final Block block, final Map<Location, SignProtec
     }
 
     private SignProtectionState checkProtectionSign(final Block block, final User user, final String username) {
-        if (block.getType() == Material.SIGN_POST || block.getType() == Material.WALL_SIGN) {
+        if (block.getType() == Material.SIGN || block.getType() == Material.WALL_SIGN) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/signs/SignProtection.java]: m a t e r i a l u t i l

File: Essentials/src/com/earth2me/essentials/storage/BukkitConstructor.java [hunk 18]
```diff
@@ -59,15 +61,9 @@ protected String constructScalarRefl(ScalarNode scalarNode) {
...
                 final String val = constructScalarRefl((ScalarNode) node);
-                Material mat;
-                if (NumberUtil.isInt(val)) {
-                    final int typeId = Integer.parseInt(val);
-                    mat = Material.getMaterial(typeId);
-                } else {
-                    mat = Material.matchMaterial(val);
-                }
-                return mat;
+                return Material.matchMaterial(val);
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/storage/BukkitConstructor.java:64]: Potentially not backwards-compatible?

File: Essentials/src/com/earth2me/essentials/utils/LocationUtil.java [hunk 19]
```diff
@@ -24,147 +22,25 @@
...
-        // Additional Materials added in by Essentials
-        HOLLOW_MATERIALS.add(Material.SEEDS);
-        HOLLOW_MATERIALS.add(Material.SIGN_POST);
-        HOLLOW_MATERIALS.add(Material.WOODEN_DOOR);
-        HOLLOW_MATERIALS.add(Material.WALL_SIGN);
-        HOLLOW_MATERIALS.add(Material.STONE_PLATE);
-        HOLLOW_MATERIALS.add(Material.IRON_DOOR_BLOCK);
-        HOLLOW_MATERIALS.add(Material.WOOD_PLATE);
-        HOLLOW_MATERIALS.add(Material.FENCE_GATE);
+        for (Material mat : Material.values()) {
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/utils/LocationUtil.java:25]: Check whether this still works on 1.12 and below.

File: Essentials/src/com/earth2me/essentials/utils/VersionUtil.java [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import org.bukkit.Bukkit;
+
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+//1.13.1-R0.1-SNAPSHOT
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/utils/VersionUtil.java]: Test line please ignore

File: Essentials/test/com/earth2me/essentials/FakeServer.java [hunk 21]
```diff
@@ -969,6 +994,31 @@ public Advancement getAdvancement(NamespacedKey key) {
         throw new UnsupportedOperationException("Not supported yet.");
     }
 
+    @Override
+    public BlockData createBlockData(Material material) {
+        return null;
```
Comment (mdcfe) [inline Essentials/test/com/earth2me/essentials/FakeServer.java:1029]: Throw exception here.

File: EssentialsProtect/src/com/earth2me/essentials/protect/EssentialsProtectBlockListener.java [hunk 22]
```diff
@@ -57,12 +57,12 @@ public void onBlockIgnite(BlockIgniteEvent event) {
     public void onBlockFromTo(final BlockFromToEvent event) {
         final Block block = event.getBlock();
 
-        if (block.getType() == Material.WATER || block.getType() == Material.STATIONARY_WATER) {
+        if (block.getType() == Material.WATER || block.getType() == Material.WATER) {
```
Comment (mdcfe) [inline EssentialsProtect/src/com/earth2me/essentials/protect/EssentialsProtectBlockListener.java]: MaterialUtil

File: nms/NMSProvider/src/net/ess3/nms/SpawnerProvider.java [hunk 23]
```diff
@@ -31,7 +31,13 @@
     public boolean tryProvider() {
         try {
             EntityType type = EntityType.CREEPER;
-            ItemStack is = setEntityType(new ItemStack(Material.MOB_SPAWNER), type);
+            Material MOB_SPAWNER;
+            try {
```
Comment (mdcfe) [inline nms/NMSProvider/src/net/ess3/nms/SpawnerProvider.java:35]: MaterialUtil

File: Essentials/src/com/earth2me/essentials/utils/MaterialUtil.java [hunk 24]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+
+import static com.earth2me.essentials.utils.EnumUtil.getMaterial;
+
+public class MaterialUtil {
+
+    private static final Set<Material> BEDS = new HashSet<>();
+
```
Comment (mdcfe) [inline Essentials/src/com/earth2me/essentials/utils/MaterialUtil.java:27]: Add static constants for SIGN_POST and SPAWNER since they're used all over the place.

**Review by mdcfe — COMMENTED** [review]
Note that most references to `MaterialUtil` actually mean `EnumUtil.getMaterial`, but I didn't realise until I was almost done.

**Comment by mibby** [comment]
@md678685 You'll need to add an alias to MooshroomSpawnEgg since it doesn't fit on a buy sign. Same with SkeletonHorseSpawnEgg, ZombieHorseSpawnEgg, and PolarBearSpawnEgg. Probably a few more somewhere as well.

I've just been manually adding aliases of MobNameEgg.

Edit: Came across quite a few extra items that needed shorter aliases and noticed quite a few items like saplings missing aliases.

**Comment by mdcfe** [comment]
@mibby I've added new aliases for all the mob eggs - is there any chance you could get a list of the aliases you've noticed are missing?

I'm aware that there are lots of stairs and slabs still missing at this point, but I haven't noticed any missing sapling aliases yet.

**Comment by reteps** [comment]
Are we good?

**Comment by mdcfe** [comment]
@drtshock This is now ready when you're happy to merge it.

File: Essentials/src/com/earth2me/essentials/commands/Commandrepair.java [hunk 25]
```diff
@@ -38,7 +39,7 @@ public void run(final Server server, final User user, final String commandLabel,
     }
 
     public void repairHand(User user) throws Exception {
-        final ItemStack item = user.getBase().getItemInHand();
+        final ItemStack item = user.getBase().getInventory().getItemInMainHand();
```
Comment (drtshock) [inline Essentials/src/com/earth2me/essentials/commands/Commandrepair.java]: Off hand was added in 1.9 so this will break on any versions before then (such as the popular 1.8.8). Do we want to add a version check for this? Or are we breaking <1.9 support completely?

**Comment by drtshock** [comment]
Other than my question about <1.9 support, looks good. Amazing work.

File: Essentials/src/com/earth2me/essentials/commands/Commandrepair.java (hunk 25 above)
Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/commands/Commandrepair.java]: Seems like `/repair`, `/more` and `/itemdb` were updated to do this in one of the initial 1.13 commits (1a820ad9b7d2564d836f3044f770aec56c0b04a1). We could implement a method that tries `getItemInMainHand` and falls back to `getItemInHand`? (I'm not a fan of continual 1.8.x workarounds, but this seems like a trivial change.)

Comment (mdcfe) [reply Essentials/src/com/earth2me/essentials/commands/Commandrepair.java]: Update: It seems `User#getItemInHand` was added to address this, but wasn't used for updating these commands.

**Review by drtshock — DISMISSED** [review]

**Review by drtshock — APPROVED** [review]

**Comment by drtshock** [comment]
WOOOOOOOO

**Comment by smmmadden** [comment]
I second that!  Congrats!!
